### PR TITLE
Merging view implementation : empty adapters

### PR DESCRIPTION
**Pull request by tadam50**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
No

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
New adaptated classes are developped. No implementation are done.
Each method adaptated throw an exception

**What is the current behavior?** *(You can also link to an open issue here)*
Merging view module implementation need to implement adapters

**What is the new behavior (if this is a feature change)?**
Merging view module implementation step : empty adapters implementation needed

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java [hunk 1]
```diff
@@ -275,422 +281,439 @@ public int getSubstationCount() {
     @Override
     public Substation getSubstation(final String id) {
         return getSubstationStream()
-                .filter(s -> id.compareTo(s.getId()) == 0)
+                .filter(s -> id.equals(s.getId()))
                 .findFirst()
                 .orElse(null);
     }
```
Comment (mathbagu) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java]: Not sure this method is efficient: why don't use the index instead?

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java [hunk 2]
```diff
@@ -275,422 +281,439 @@ public int getSubstationCount() {
...
     @Override
     public int getVoltageLevelCount() {
-        throw NOT_IMPLEMENTED_EXCEPTION;
+        return index.getVoltageLevels().size();
     }
 
     @Override
     public VoltageLevel getVoltageLevel(final String id) {
-        throw NOT_IMPLEMENTED_EXCEPTION;
+        return getVoltageLevelStream().filter(s -> id.equals(s.getId())).findFirst().orElse(null);
```
Comment (mathbagu) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java]: Same remark here

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java [hunk 3]
```diff
@@ -275,422 +281,439 @@ public int getSubstationCount() {
...
-        throw NOT_IMPLEMENTED_EXCEPTION;
+    public int getBatteryCount() {
+        return index.getBatteries().size();
     }
 
     @Override
-    public int getBranchCount() {
-        throw NOT_IMPLEMENTED_EXCEPTION;
+    public Battery getBattery(final String id) {
+        return getBatteryStream().filter(s -> id.equals(s.getId())).findFirst().orElse(null);
```
Comment (mathbagu) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java]: Same remark here

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java [hunk 4]
```diff
@@ -275,422 +281,439 @@ public int getSubstationCount() {
...
-        throw NOT_IMPLEMENTED_EXCEPTION;
+    public int getVscConverterStationCount() {
+        return index.getVscConverterStations().size();
     }
 
     @Override
-    public TieLineAdder newTieLine() {
-        throw NOT_IMPLEMENTED_EXCEPTION;
+    public VscConverterStation getVscConverterStation(final String id) {
+        return getVscConverterStationStream().filter(s -> id.equals(s.getId())).findFirst().orElse(null);
```
Comment (mathbagu) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java]: Same remark here

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java [hunk 5]
```diff
@@ -275,422 +281,439 @@ public int getSubstationCount() {
...
     @Override
     public int getTwoWindingsTransformerCount() {
-        throw NOT_IMPLEMENTED_EXCEPTION;
+        return index.getTwoWindingsTransformers().size();
     }
 
     @Override
     public TwoWindingsTransformer getTwoWindingsTransformer(final String id) {
-        throw NOT_IMPLEMENTED_EXCEPTION;
+        return getTwoWindingsTransformerStream().filter(s -> id.equals(s.getId())).findFirst().orElse(null);
```
Comment (mathbagu) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java]: Same remark here

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java [hunk 6]
```diff
@@ -275,422 +281,439 @@ public int getSubstationCount() {
...
     }
 
+    // Switches
     @Override
-    public Iterable<ThreeWindingsTransformer> getThreeWindingsTransformers() {
-        throw NOT_IMPLEMENTED_EXCEPTION;
+    public Switch getSwitch(final String id) {
+        return getSwitchStream().filter(s -> id.equals(s.getId())).findFirst().orElse(null);
```
Comment (mathbagu) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java]: Same remark here

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java [hunk 7]
```diff
@@ -275,422 +281,439 @@ public int getSubstationCount() {
...
-        throw NOT_IMPLEMENTED_EXCEPTION;
+    public int getStaticVarCompensatorCount() {
+        return index.getStaticVarCompensators().size();
     }
 
     @Override
-    public Generator getGenerator(final String id) {
-        throw NOT_IMPLEMENTED_EXCEPTION;
+    public StaticVarCompensator getStaticVarCompensator(final String id) {
+        return getStaticVarCompensatorStream().filter(s -> id.equals(s.getId())).findFirst().orElse(null);
```
Comment (mathbagu) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java]: Same remark

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java [hunk 8]
```diff
@@ -275,422 +281,439 @@ public int getSubstationCount() {
...
-        throw NOT_IMPLEMENTED_EXCEPTION;
+    public int getShuntCompensatorCount() {
+        return index.getShuntCompensators().size();
     }
 
     @Override
-    public Battery getBattery(final String id) {
-        throw NOT_IMPLEMENTED_EXCEPTION;
+    public ShuntCompensator getShuntCompensator(final String id) {
+        return getShuntCompensatorStream().filter(s -> id.equals(s.getId())).findFirst().orElse(null);
```
Comment (mathbagu) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java]: Same remark

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java [hunk 9]
```diff
@@ -275,422 +281,439 @@ public int getSubstationCount() {
...
     @Override
     public int getLoadCount() {
-        throw NOT_IMPLEMENTED_EXCEPTION;
+        return index.getLoads().size();
     }
 
     @Override
     public Load getLoad(final String id) {
-        throw NOT_IMPLEMENTED_EXCEPTION;
+        return getLoadStream().filter(s -> id.equals(s.getId())).findFirst().orElse(null);
```
Comment (mathbagu) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java]: Same remark

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java [hunk 10]
```diff
@@ -275,422 +281,439 @@ public int getSubstationCount() {
...
+    public Generator getGenerator(final String id) {
+        return getGeneratorStream().filter(s -> id.equals(s.getId())).findFirst().orElse(null);
     }
 
+    // BusbarSections
     @Override
-    public Iterable<DanglingLine> getDanglingLines() {
-        throw NOT_IMPLEMENTED_EXCEPTION;
+    public BusbarSection getBusbarSection(final String id) {
+        return getBusbarSectionStream().filter(s -> id.equals(s.getId())).findFirst().orElse(null);
```
Comment (mathbagu) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java]: Same remark

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java [hunk 11]
```diff
@@ -275,422 +281,439 @@ public int getSubstationCount() {
...
-        throw NOT_IMPLEMENTED_EXCEPTION;
+    public int getLccConverterStationCount() {
+        return index.getLccConverterStations().size();
     }
 
     @Override
-    public StaticVarCompensator getStaticVarCompensator(final String id) {
-        throw NOT_IMPLEMENTED_EXCEPTION;
+    public LccConverterStation getLccConverterStation(final String id) {
+        return getLccConverterStationStream().filter(s -> id.equals(s.getId())).findFirst().orElse(null);
```
Comment (mathbagu) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java]: Same remark

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java [hunk 12]
```diff
@@ -275,422 +281,439 @@ public int getSubstationCount() {
...
-        throw NOT_IMPLEMENTED_EXCEPTION;
+    public int getHvdcConverterStationCount() {
+        return index.getHvdcConverterStations().size();
     }
 
     @Override
-    public int getSwitchCount() {
-        throw NOT_IMPLEMENTED_EXCEPTION;
+    public HvdcConverterStation<?> getHvdcConverterStation(final String id) {
+        return getHvdcConverterStationStream().filter(s -> id.equals(s.getId())).findFirst().orElse(null);
```
Comment (mathbagu) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java]: Same remark

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java [hunk 13]
```diff
@@ -275,422 +281,439 @@ public int getSubstationCount() {
...
     }
 
+    // Branches
     @Override
-    public BusbarSection getBusbarSection(final String id) {
-        throw NOT_IMPLEMENTED_EXCEPTION;
+    public Branch getBranch(final String id) {
+        return getBranchStream().filter(s -> id.equals(s.getId())).findFirst().orElse(null);
```
Comment (mathbagu) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java]: Same remark

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java [hunk 14]
```diff
@@ -275,422 +281,439 @@ public int getSubstationCount() {
...
-        throw NOT_IMPLEMENTED_EXCEPTION;
+    public int getThreeWindingsTransformerCount() {
+        return index.getThreeWindingsTransformers().size();
     }
 
     @Override
-    public HvdcConverterStation<?> getHvdcConverterStation(final String id) {
-        throw NOT_IMPLEMENTED_EXCEPTION;
+    public ThreeWindingsTransformer getThreeWindingsTransformer(final String id) {
+        return getThreeWindingsTransformerStream().filter(s -> id.equals(s.getId())).findFirst().orElse(null);
```
Comment (mathbagu) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java]: Same remark

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java [hunk 15]
```diff
@@ -275,422 +281,439 @@ public int getSubstationCount() {
...
-        throw NOT_IMPLEMENTED_EXCEPTION;
+    public int getLineCount() {
+        return index.getLines().size();
     }
 
     @Override
-    public LccConverterStation getLccConverterStation(final String id) {
-        throw NOT_IMPLEMENTED_EXCEPTION;
+    public Line getLine(final String id) {
+        return getLineStream().filter(s -> id.equals(s.getId())).findFirst().orElse(null);
```
Comment (mathbagu) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java]: Same remark

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java [hunk 16]
```diff
@@ -275,422 +281,439 @@ public int getSubstationCount() {
...
-        throw NOT_IMPLEMENTED_EXCEPTION;
+    public int getDanglingLineCount() {
+        return index.getDanglingLines().size();
     }
 
     @Override
-    public VscConverterStation getVscConverterStation(final String id) {
-        throw NOT_IMPLEMENTED_EXCEPTION;
+    public DanglingLine getDanglingLine(final String id) {
+        return getDanglingLineStream().filter(s -> id.equals(s.getId())).findFirst().orElse(null);
```
Comment (mathbagu) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java]: Same remark

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingViewIndex.java [hunk 17]
```diff
@@ -15,18 +15,49 @@
...
-    private final Map<Identifiable<?>, AbstractAdapter<?>> adaptersCacheMap = new WeakHashMap<>();
+    private final Map<Identifiable<?>, AbstractAdapter<?>> identifiableAdaptersCached = new WeakHashMap<>();
+
+    private final Map<Component, AbstractAdapter<?>> componentAdaptersCached = new WeakHashMap<>();
+
+    private final Map<Terminal, AbstractAdapter<?>> terminalAdaptersCached = new WeakHashMap<>();
+
+    private final Map<PhaseTapChanger, AbstractAdapter<?>> ptcAdaptersCached = new WeakHashMap<>();
+
+    private final Map<RatioTapChanger, AbstractAdapter<?>> rtcAdaptersCached = new WeakHashMap<>();
```
Comment (mathbagu) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingViewIndex.java]: You can probably remove the "adapters" suffix from the name of all these variables. This information is carried by the type of the map.

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingViewIndex.java [hunk 18]
```diff
@@ -85,26 +124,321 @@ void checkAndAdd(final Network other) {
...
+        if (b == null) {
+            return null;
+        }
+        switch (b.getType()) {
+            case LINE:
+                return getLine((Line) b);
+            case TWO_WINDINGS_TRANSFORMER:
+                return getTwoWindingsTransformer((TwoWindingsTransformer) b);
+            default:
+                throw new AssertionError(b.getType().name() + " is not valid to be immutablized to branch");
```
Comment (mathbagu) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingViewIndex.java]: Same remark here

**Comment by mathbagu** [comment]
@tadam50 could you have a look to the sonar report: there is a lot of smells, maybe you can fix some of them.
About duplicates, you can probably introduce an AbstractInjectionAdapter (and the Branch equivalent...): fitting the hierarchy will help

File: iidm/iidm-mergingview/src/test/java/com/powsybl/iidm/mergingview/TieLineAdapterTest.java [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+        TestUtil.notImplemented(tieLine::getB1);
+        TestUtil.notImplemented(() -> tieLine.setB1(0.0d));
+        TestUtil.notImplemented(tieLine::getB2);
+        TestUtil.notImplemented(() -> tieLine.setB2(0.0d));
+        TestUtil.notImplemented(tieLine::getUcteXnodeCode);
+        TestUtil.notImplemented(tieLine::getHalf1);
+        TestUtil.notImplemented(tieLine::getHalf2);
+    }
+
+    private Network createNetwork() {
```
[the remark this answers is not in the document]
Comment (tadam50) [reply iidm/iidm-mergingview/src/test/java/com/powsybl/iidm/mergingview/TieLineAdapterTest.java]: static keyword added

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java (hunk 1 above)
Comment (tadam50) [reply iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java]: yes, i changed in order to use NetworkIndex from each Network

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingViewIndex.java (hunk 17 above)
Comment (tadam50) [reply iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingViewIndex.java]: Done

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingViewIndex.java [hunk 20]
```diff
@@ -85,26 +124,321 @@ void checkAndAdd(final Network other) {
...
+            case SHUNT_COMPENSATOR:
+                return getShuntCompensator((ShuntCompensator) connectable);
+            case DANGLING_LINE:
+                return getDanglingLine((DanglingLine) connectable);
+            case STATIC_VAR_COMPENSATOR:
+                return getStaticVarCompensator((StaticVarCompensator) connectable);
+            case HVDC_CONVERTER_STATION:
+                return getHvdcConverterStation((HvdcConverterStation) connectable);
+            default:
+                throw new AssertionError(connectable.getType().name() + " is not valid to be immutablized to connectable");
```
[the remark this answers is not in the document]
Comment (tadam50) [reply iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingViewIndex.java]: Yes, i changed that

**Comment by tadam50** [comment]
> A remark: will the implementations of for example `Line` and `TwoWindingsTransformer` 's method `getTerminal1()` be very different or will they be identical? Just asking to know if there will be a `AbstractBranchAdapter` later for refactoring.

**Comment by tadam50** [comment]
@mathbagu 
About duplicates, i introduce an AbstractInjectionAdapter, AbstractBranchAdapter, AbstractTapChnagerAdapter & AbstractHvdcConverterStationAdapter

**Pull request merged**